verilog/preprocessor/verilog_preprocess.cc

```diff
@@ -587,48 +601,18 @@ absl::Status VerilogPreprocess::HandleInclude(
   std::filesystem::path file_path =
       std::string(token_text.substr(1, token_text.size() - 2));
 
-  std::string file_path_to_search = file_path.string();
-
-  if (file_path.is_relative()) {
-    bool found = false;
-    for (const auto& base_dir : preprocess_info_.include_dirs) {
-      file_path_to_search =
-          verible::file::JoinPath(base_dir, file_path.string());
-      auto search_status = verible::file::FileExists(file_path_to_search);
-      if (search_status.ok()) {
-        found = true;
-        break;
-      }
-    }
-    if (!found) {
-      preprocess_data_.errors.push_back(
-          {**token_iter,
-           absl::StrCat(file_path_to_search, ": no such file found.")});
-      return absl::InvalidArgumentError("ERROR: included file can't be found.");
-    }
-  } else {
-    if (auto status = verible::file::FileExists(file_path_to_search);
-        !status.ok()) {
-      preprocess_data_.errors.push_back(
-          {**token_iter, std::string(status.message())});
-      return status;
-    }
-  }
-
-  // Actually open the file.
-  std::string source_contents;
-  if (auto status =
-          verible::file::GetContents(file_path_to_search, &source_contents);
-      !status.ok()) {
+  const auto status_or_file = project_->OpenIncludedFile(file_path.string());
```

Comment:
Intead of using `VerilogProject` (which results in all the cyclic dependencies) maybe you can pass in a 
`std::function<absl::StatusOr<VerilogSourceFile*>(absl::string_view filename)> file_pener', which you then call here.

Whoever calls HandleInclude() then has to pass the function that can be a wrap around VerilogProject.

(Of course, longer term, we don't really want a `VerilogSourceFile` I suspect, but just the underlying file `std::shared_ptr<MemBlock>` ?)

Reply:
I tried the following:
```c++
  // verilog_preprocess.h
  using FileOpener = std::function<absl::StatusOr<VerilogSourceFile>(absl::string_view)>;
  FileOpener file_opener_ = nullptr;
```


```c++
  // verilog_preprocess.cc
  const auto status_or_file = file_opener_(file_path.string());
  /* const auto status_or_file = project_->OpenIncludedFile(file_path.string()); */
  if (!status_or_file.ok()) {
    preprocess_data_.errors.push_back(
        {**token_iter, std::string(status_or_file.status().message())});
    return status_or_file.status();
  }
  const auto source_contents = status_or_file->GetContent()->AsStringView();
```
To get the `source_contents` I would need to call `GetContent()` from `VerilogSourceFile`, that's why I should keep `#include "verilog_project.c"` in `verilog_preprocess.cc`, so the issue remains.

Am I missing something?

Reply:
what if your file_opener returns an `absl::string_view`, not a super-specific `VerilogSourceFile` ?

```
using FileOpener = std::function<absl::StatusOr<absl::string_view>(absl::string_view filename)>;
```

And an implementation can be roughly something like this:
```
 FileOpener my_fun = [project](absl::string_view filename) -> absl::StatusOr<absl::string_view {
    auto result = project->OpenTheFile(filename);
    if (!result.status().ok()) return result.status();
    return result->GetContent();
 }
```

Reply:
I think that's also won't fit in, because the implementation is needed inside `HandleInclude()` to open the files and there can't be `VerilogProject project` in there to capture it in `my_fun`, right?

Reply:
Yes, but in this case the caller is HandleInclude() itself.

Reply:
It will be the caller relative to the new VerilogPreprocess which will scan the contents of the included file.